bouncer/views.py

```diff
@@ -25,11 +29,16 @@ def annotation(self):
                 id=self.request.matchdict["id"])
         except exceptions.NotFoundError:
             statsd.incr("views.annotation.404.annotation_not_found")
-            raise httpexceptions.HTTPNotFound()
+            raise httpexceptions.HTTPNotFound("Annotation not found")
 
         annotation_id = document["_id"]
         document_uri = document["_source"]["uri"]
 
+        if not (document_uri.startswith("http://") or
+                document_uri.startswith("https://")):
+            raise httpexceptions.HTTPUnprocessableEntity(
+                _("We can only link to http(s) URLs"))
```

Comment:
After spending some quality time sifting through all the special ways we store annotations at the moment, I found quite a few that don't have a `uri`, and even some that don't have anything that could be used as a URI. 
The Postgres migration code for example tries to read the `uri` first, and then falls back to `target.link[0]`, and if that one doesn't exist, refuses to migrate the annotation.

My suggestion for bouncer is something like:

1. Read from `uri`
2. When `uri` is missing or not a http(s) uri, find the first http(s) uri from `target.link` with `rel == 'canonical'` (careful, sometimes `target.link` is not a list but a string)
3. When that one is missing as well, show an error.